File: go/kan/codensity.go
package capitan

type Codensity struct {
    Value func(func(interface{}) interface{}) interface{}
}

func Return(x interface{}) Codensity {
    return Codensity{func(k func(interface{}) interface{}) interface{} {
        return k(x)
    }}
}

func (m Codensity) Bind(f func(interface{}) Codensity) Codensity {
    return Codensity{func(k func(interface{}) interface{}) interface{} {
        return m.Value(func(x interface{}) interface{} {
            return f(x).Value(k)
        })
    }}
}

func (m Codensity) Map(f func(interface{}) interface{}) Codensity {
    return Codensity{func(k func(interface{}) interface{}) interface{} {
        return m.Value(func(x interface{}) interface{} {
            return k(f(x))
        })
    }}
}

func LowerCodensity(m Codensity) interface{} {
    return m.Value(func(x interface{}) interface{} { return x })
}

func Reset(m Codensity) interface{} {
    return LowerCodensity(m)
}

func Shift(f func(func(interface{}) Codensity) Codensity) Codensity {
    aux := func(k func(interface{}) interface{}) interface{} {
        return f(func(x interface{}) Codensity {
            return Return(x)
        }).Value(k)
    }
    return Codensity{func(k func(interface{}) interface{}) interface{} {
        return aux(k)
    }}
}

File: rust/src/kan/codensity.rs
struct Codensity<'a, M, A>(Box<dyn 'a + Fn(Box<dyn 'a + Fn(A) -> M<B>>) -> M<B>>);

impl<'a, M, A> Codensity<'a, M, A>
where
    M: Monad<A>,
{
    fn return_(x: A) -> Codensity<'a, M, A> {
        Codensity(Box::new(move |k| k(x)))
    }

    fn bind<B, F>(self, f: F) -> Codensity<'a, M, B>
    where
        F: 'a + Fn(A) -> Codensity<'a, M, B>,
    {
        Codensity(Box::new(move |k| {
            (self.0)(Box::new(move |a| {
                let codensity = f(a);
                (codensity.0)(k)
            }))
        }))
    }

    fn map<B, F>(self, f: F) -> Codensity<'a, M, B>
    where
        F: 'a + Fn(A) -> B,
    {
        Codensity(Box::new(move |k| {
            (self.0)(Box::new(move |a| k(f(a))))
        }))
    }

    fn lower_codensity(self) -> M<A> {
        (self.0)(Box::new(|a| M::return_(a)))
    }
}

fn reset<'a, M, A>(m: Codensity<'a, M, A>) -> M<A>
where
    M: Monad<A>,
{
    m.lower_codensity()
}

fn shift<'a, M, A, B>(f: impl 'a + Fn(Box<dyn 'a + Fn(A) -> Codensity<'a, M, B>>) -> Codensity<'a, M, A>) -> Codensity<'a, M, A>
where
    M: Monad<A>,
    M: Monad<B>,
{
    let aux = move |k| {
        let codensity = f(Box::new(move |a| Codensity::return_(a)));
        (codensity.0)(k)
    };
    Codensity(Box::new(move |k| aux(k)))
}